src/keymap.h

```diff
@@ -387,6 +387,30 @@ struct xkb_mod_set {
     xkb_mod_mask_t explicit_vmods;
 };
 
+/*
+ * Our current implementation with continuous arrays does not allow efficient
+ * mapping of keycodes, let’s be further conservative.
+ */
+#define XKB_KEYCODE_MAX_IMPL 0xfff
```

Comment:
It should be possible to use a typed const instead of define, always seems better to me. The other consts in this file are defines but we can start with this one.

I also suggest using a decimal here, it's easier to understand the limit than in hex.

```suggestion
const xkb_keycode_t XKB_KEYCODE_MAX_IMPL = 4095;
```

Reply:
But it will not work with `static_assert`? We would need C23’s `constexpr` or using `extern` in the header + usual `assert`.

As for the hex format: no strong opinion here, I used it because it’s the format of Linux (upper) keycodes.